JSM: Added module and TS file for DRACOLoader

soadzoor opened the pull request.

soadzoor committed:
JSM: Added module and TS file for DRACOLoader

soadzoor committed:
Added changes to the docs

Mugen87 commented on examples/jsm/loaders/DRACOLoader.d.ts:
```diff
@@ -0,0 +1,9 @@
+
+export class DRACOLoader {
+  constructor(manager?: LoadingManager);
+
```
You should also add the actual public interface of the loader. So methods like `setPath()` or `setVerbosity()`.

soadzoor replied:
@Mugen87 How could I know which of these methods are meant to be public and which aren't? In the examples I've found, only a static `setDecoderPath` is used.

gkjohnson replied:
You can see the public API in the docs:

https://threejs.org/docs/#examples/loaders/DRACOLoader

It looks like `setVerbosity` isn't documented, though.

Mugen87 replied:
It can't hurt to provide typings for the entire public accessible interface.

gkjohnson replied:
> It can't hurt to provide typings for the entire public accessible interface.

 Oh I agree -- it's just not necessarily clear what's considered "public" if it's not in the docs.

donmccurdy replied:
I would suggest skipping the following methods, at least –

- decodeDracoFileInternal
- addAttributeToGeometry
- convertDracoGeometryTo3JS
- getAttributeOptions

Mugen87 commented:
You should also update [modularize.js](https://github.com/mrdoob/three.js/blob/dev/utils/modularize.js) since I hope you have converted the module with the script (and not manually).

soadzoor commented:
> You should also update [modularize.js](https://github.com/mrdoob/three.js/blob/dev/utils/modularize.js) since I hope you have converted the file with the script (and not manually).

Lol, actually I did it manually. Which script does it automatically? I tried each npm scripts from the `package.json` file, non of them seems to be doing any files to the jsm folder. I might be wrong though.

soadzoor committed:
Added DRACOLoader to modularize.js

gkjohnson commented:
> An npm script would be nice though.

What do you mean an npm script? The conversion for the THREE examples is very specialized so I'm not sure if a published script would make sense.

Also did you regenerate the module using the script?

Mugen87 commented:
> The conversion for the THREE examples is very specialized so I'm not sure if a published script would make sense.

Yeah, a `npm` script is not necessary here.

soadzoor commented:
> > An npm script would be nice though.
> 
> What do you mean an npm script? The conversion for the THREE examples is very specialized so I'm not sure if a published script would make sense.
> 
> Also did you regenerate the module using the script?

I did, but I haven't pushed yet, because it fails on my tests. For some reason, the "imports" are not generated. Any idea on that?

gkjohnson commented:
> I did, but I haven't pushed yet, because it fails on my tests. For some reason, the "imports" are not generated. Any idea on that?

You'll have to specify the dependency files manually, like in the `Line2.js` case:

```js
{ path: 'lines/Line2.js', dependencies: [ { name: 'LineSegments2', path: 'lines/LineSegments2.js' }, { name: 'LineGeometry', path: 'lines/LineGeometry.js' }, { name: 'LineMaterial', path: 'lines/LineMaterial.js' } ], ignoreList: [] }
```

soadzoor commented:
> > I did, but I haven't pushed yet, because it fails on my tests. For some reason, the "imports" are not generated. Any idea on that?
> 
> You'll have to specify the dependency files manually, like in the `Line2.js` case:
> 
> ```js
> { path: 'lines/Line2.js', dependencies: [ { name: 'LineSegments2', path: 'lines/LineSegments2.js' }, { name: 'LineGeometry', path: 'lines/LineGeometry.js' }, { name: 'LineMaterial', path: 'lines/LineMaterial.js' } ], ignoreList: [] }
> ```

I was thinking the same, but then realized that for example `DDSLoader` doesn't have those dependency files specified, yet the generated file has all the needed imports included. And not just the DDSLoader, but a lot of other examples. How does this work?

gkjohnson commented:
The script is pretty simple so you can look through it to see the details of it all but only dependencies that are from the `examples` folder need to be specified in that list.

soadzoor commented:
> The script is pretty simple so you can look through it to see the details of it all but only dependencies that are from the `examples` folder need to be specified in that list.

That seems to be logical, but eg.: if OrbitControls imports `MOUSE`, which is in the core (`three.module.js`), then why can't DRACOLoader import `TrianglesDrawMode`, which is in the core. This is just an example. All of its dependencies are in the core, yet none of them are included for some reason.

donmccurdy commented:
This probably means that the DRACOLoader class is declared in a way that isn't quite compatible with the structure/order that `modularize.js` is looking for, so `modularize.js` can't tell where to inject the imports... you can try comparing the original file to other files that do work with automatic conversion, or perhaps post the output you're getting now.

Mugen87 commented:
@soadzoor Just replace the original header with this:
```
/** Copyright 2016 The Draco Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```
And also add `LoadingManager` to the ignore list. Besides, it would also be good to apply the `three.js` code style to `DracoLoader`. If you have a correctly configured IDE, this can be done with a single open/save of the original file.

soadzoor committed:
Applying three.js code style + replaced manually created jsm file with generated one

soadzoor committed:
Added public functions to d.ts

soadzoor committed:
Fixes for imports

soadzoor commented:
> @soadzoor Just replace the original header with this:
> 
> ```
> /** Copyright 2016 The Draco Authors.
>  *
>  * Licensed under the Apache License, Version 2.0 (the "License");
>  * you may not use this file except in compliance with the License.
>  * You may obtain a copy of the License at
>  *
>  *      http://www.apache.org/licenses/LICENSE-2.0
>  *
>  * Unless required by applicable law or agreed to in writing, software
>  * distributed under the License is distributed on an "AS IS" BASIS,
>  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
>  * See the License for the specific language governing permissions and
>  * limitations under the License.
>  */
> ```
> 
> And also add `LoadingManager` to the ignore list. Then rerun the script. Besides, it would also be good to apply the `three.js` code style to `DracoLoader`. If you have a correctly configured IDE, this can be done with a single open/save of the original file.

Replacing the license with this format actually solved the `import` issue, thanks a lot! I keep wondering why that is though, it's very strange. 😮 

I applied the three.js coding style to the code.

I didn't add `LoadingManager` to the ignore list though, can you explain why I need to do that? It's not on the ignore list for other loaders either.

I added functions to the d.ts file: all that can be found in the docs is now added, and even some that are not in the docs. But for some, I've found that they're used internally, and it didn't really make sense to me to make them public, like the ones @donmccurdy mentioned.

Mugen87 commented:
> I didn't add LoadingManager to the ignore list though, can you explain why I need to do that?

It seems you are not working with a properly configured IDE/editor, otherwise you would immediately see that the import `LoadingManager` is unnecessary. The import is still generated because `THREE.LoadingManager` is noted in a comment. `modularize.js` can't detect this circumstance.

> I keep wondering why that is though, it's very strange. 😮

When you carefully read the code in `modularize.js`, you will see that the regex for the `_IMPORTS_` directive is not compatible with the previous comment in `DracoLoader`. It's easier to adjust the file than enhancing the script though. `modularize.js` is intended for library developers. It's not important that the script is user-friendly.

Mugen87 commented on examples/jsm/loaders/DRACOLoader.d.ts:
```diff
@@ -0,0 +1,21 @@
+  setVerbosity(level: number): DRACOLoader;
+  setDrawMode(drawMode: TrianglesDrawModes): DRACOLoader;
+  setSkipDequantization(attributeName: 'position', skip?: boolean): DRACOLoader;
+  isVersionSupported(version: number, callback: (isVersionSupported: boolean) => any);
```
> isVersionSupported(version: number, callback: (isVersionSupported: boolean) => any);

The return type (`void`) is missing.

soadzoor replied:
Nice catch, thanks. Fixed now.

soadzoor committed:
Added 'LoadingManager' to ignoreList + added `void` return type for isVersionSupported method

soadzoor commented:
> > I didn't add LoadingManager to the ignore list though, can you explain why I need to do that?
> 
> It seems you are not working with a properly configured IDE/editor, otherwise you would immediately see that the import of `LoadingManager` is unnecessary. The import is still generated because `THREE.LoadingManager` is noted in a comment. `modularize.js` can't detect this circumstance.
> 
> > I keep wondering why that is though, it's very strange. 😮
> 
> When you carefully read the code in `modularize.js`, you will see that the regex for the `_IMPORTS_` directive is not compatible with the previous comment in `DracoLoader`. It's easier to adjust the file than enhancing the script though. `modularize.js` is intended for library developers. It's not important that the script is user-friendly.

Thanks, I updated the files based on this

Mugen87 commented on examples/js/loaders/DRACOLoader.js:
```diff
@@ -1,411 +1,580 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+"use strict";
```
Sorry, one last thing^^:

'use strict'; has no effect in modules and we generally do not use it in loaders. So this line can be removed.

mrdoob commented:
Thanks!

soadzoor commented:
@mrdoob Can you please add this to the latest changelog (r106)?

Mugen87 commented:
You mean the [release notes](https://github.com/mrdoob/three.js/releases/tag/r106)? It's part of the point: `Added module and TS files for more examples code.`

It would be too effortful to mentioned each single file since there were so many of them. Hence, all JSM changes are bundled in a single note.

Mugen87 commented:
Done (see section `Added module and TS files for more examples code.`)
